src/core/src/op/convert_align_types.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "openvino/op/convert_align_types.hpp"
+
+#include "itt.hpp"
+#include "validation_util.hpp"
+
+namespace ov {
+namespace op {
+namespace {
+
+std::unordered_map<size_t, element::Type> bit_to_int{
+    {4, element::i4},
+    {8, element::i8},
+    {16, element::i16},
+    {32, element::i32},
+    {64, element::i64},
+};
+
+element::Type infer_types(const v14::ConvertAlignTypes* op) {
+    const auto lhs = op->input(0);
+    const auto rhs = op->input(1);
+    const auto promote_unsafe = op->get_promote_unsafe();
+    const auto pytorch_scalar_align = op->get_pytorch_scalar_align();
+    const auto u64_promotion_target = op->get_u64_integer_promotion_target();
+    const auto supported_types = {element::dynamic,
+                                  element::boolean,
+                                  element::f16,
+                                  element::f32,
+                                  element::f64,
+                                  element::i4,
+                                  element::i8,
+                                  element::i16,
+                                  element::i32,
+                                  element::i64,
+                                  element::u1,
+                                  element::u4,
+                                  element::u8,
+                                  element::u16,
+                                  element::u32,
+                                  element::u64,
+                                  element::f8e4m3,
+                                  element::f8e5m2,
+                                  element::bf16};
+    const auto& lhs_type = lhs.get_element_type();
+    const auto& rhs_type = rhs.get_element_type();
+    NODE_VALIDATION_CHECK(op,
+                          std::find(supported_types.begin(), supported_types.end(), lhs_type) != supported_types.end());
+    NODE_VALIDATION_CHECK(op,
+                          std::find(supported_types.begin(), supported_types.end(), rhs_type) != supported_types.end());
+    if (lhs_type.is_dynamic() || rhs_type.is_dynamic()) {
+        return element::dynamic;
+    }
+    if (lhs_type == rhs_type)
+        return lhs_type;
+    if (lhs_type == element::boolean)
+        return rhs_type;
+    if (rhs_type == element::boolean)
+        return lhs_type;
+    const auto& lhs_rank = lhs.get_partial_shape().rank();
+    const auto& rhs_rank = rhs.get_partial_shape().rank();
+    const bool is_lhs_scalar = lhs_rank.is_static() ? lhs_rank.get_length() == 0 : false;
```

Comment:
```suggestion
    const auto is_lhs_scalar = lhs_rank.is_static() && lhs_rank.get_length() == 0;
```
make helper to check scalar shape, maybe there is some helper function exists?